clippy_lints/src/methods/implicit_clone.rs

```diff
@@ -0,0 +1,37 @@
+use crate::utils::span_lint_and_sugg;
+use if_chain::if_chain;
+use rustc_errors::Applicability;
+use rustc_hir as hir;
+use rustc_hir::ExprKind;
+use rustc_lint::LateContext;
+use rustc_middle::ty::TyS;
+use rustc_span::symbol::Symbol;
+
+use super::IMPLICIT_CLONE;
+use clippy_utils::{is_expr_container_diagnostic_item, is_type_diagnostic_item};
+
+pub fn check(
+    cx: &LateContext<'_>,
+    expr: &hir::Expr<'_>,
+    trait_diagnostic: Symbol,
+    expected_type_diagnostic: Option<Symbol>,
+) {
+    if_chain! {
+        if let ExprKind::MethodCall(method_path, _, [arg], _) = &expr.kind;
+        let return_type = cx.typeck_results().expr_ty(&expr);
+        let input_type = cx.typeck_results().expr_ty(arg).peel_refs();
+        if let Some(ty_name) = input_type.ty_adt_def().map(|adt_def| cx.tcx.item_name(adt_def.did));
+        if TyS::same_type(return_type, input_type);
+        if is_expr_container_diagnostic_item(cx, expr, trait_diagnostic) &&
+            expected_type_diagnostic.map_or(true, |diagnostic| is_type_diagnostic_item(cx,return_type,diagnostic));
```

Comment:
Actually, I think it might be. The idea behind this was to prevent something like this from triggering the lint:

```
struct NotAVec {…};

impl NotAVec {
  fn to_vec(self) -> NotAVec { … }
}
```

But now that I am checking where `to_vec` is implemented (is it on slice?), this to_vec won't work because it is on `NotAVec` so there's no need to check the type.

Reply:
Right. So long as you keep the `same_type` check right above.